README.md
# ukui-display-switch
This component is to manage the extension screen.


debian/changelog
ukui-display-switch (1.0.0-1+1127.1) v101; urgency=medium

  * Initial release.

 -- zhuyunpeng <zhuyunpeng@kylinos.cn>  Fri, 27 Nov 2020 13:52:00 +0800


manager/main.cpp
/* main.cpp
 * Copyright (C) 2018 Tianjin KYLIN Information Technology Co., Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301, USA.
**/
#include <stdio.h>
#include <string>
#include <QApplication>
#include <QTextCodec>
#include <QResource>
#include <QTranslator>
#include <QLocale>
#include <QDateTime>
#include <QDebug>
#include <QStandardPaths>
#include <QProcess>
#include <QScreen>
#include <QDesktopWidget>
#include <QX11Info>
#include <X11/Xlib.h>
#include <X11/XKBlib.h>
#include <X11/keysym.h>
#include <X11/cursorfont.h>
#include <X11/extensions/Xrandr.h>
#include "display-switch/displayswitch.h"
#include "xeventmonitor.h"
#include <QFile>
#include <QtConcurrent/QtConcurrentRun>

void outputMessage(QtMsgType type, const QMessageLogContext &context, const QString &msg)
{
    QFile file("/tmp/ukui-greeter.log");
    file.open(QIODevice::WriteOnly|QIODevice::Append);
    Q_UNUSED(context)
    QDateTime dateTime = QDateTime::currentDateTime();
    QByteArray time = QString("[%1] ").arg(dateTime.toString("MM-dd hh:mm:ss.zzz")).toLocal8Bit();
    QByteArray localMsg = msg.toLocal8Bit();
    file.write(localMsg);
    file.write("\n");
    file.close();
    switch(type) {
    case QtDebugMsg:

        fprintf(stderr, "%s Debug: %s:%u: %s\n", time.constData(), context.file, context.line, localMsg.constData());
        break;
#if QT_VERSION >= QT_VERSION_CHECK(5, 5, 0)
    case QtInfoMsg:
        fprintf(stderr, "%s Info: %s:%u: %s\n", time.constData(), context.file, context.line, localMsg.constData());
        break;
#endif
    case QtWarningMsg:
        fprintf(stderr, "%s Warnning: %s:%u: %s\n", time.constData(), context.file, context.line, localMsg.constData());
        break;
    case QtCriticalMsg:
        fprintf(stderr, "%s Critical: %s:%u: %s\n", time.constData(), context.file, context.line, localMsg.constData());
        break;
    case QtFatalMsg:
        fprintf(stderr, "%s Fatal: %s:%u: %s\n", time.constData(), context.file, context.line, localMsg.constData());
        abort();
    }

}

void waitMonitorsReady()
{
    int n;

    QScreen *screen = QApplication::primaryScreen();
    while(n <= 0 && screen->geometry().x() <= 0 && screen->geometry().y() <= 0
          && screen->geometry().width() <= 0 && screen->geometry().height() <= 0){
        XRRGetMonitors(QX11Info::display(), QX11Info::appRootWindow(), false, &n);
        if(n == -1)
            qWarning() << "info: [main.cpp][waitMonitorsReady]: get monitors failed";
        else if(n > 0){

        }
        //启动xrandr，打开视频输出， 自动设置最佳分辨率
        QProcess enableMonitors;
        enableMonitors.start("xrandr --auto");
        enableMonitors.waitForFinished(-1);
    }
}

void x11_get_screen_size(int *width,int *height)
{
    Display* display;

    display = XOpenDisplay(NULL);
    if (display == NULL) {
        fprintf(stderr, "Cannot connect to X server %s/n", "simey:0");
        exit (-1);
    }
    int screen_num;

    screen_num = DefaultScreen(display);

    *width = DisplayWidth(display, screen_num);
    *height = DisplayHeight(display, screen_num);
    XCloseDisplay(display);

}

int main(int argc, char *argv[])
{
    //qInstallMessageHandler(outputMessage);

    int width = 0, height = 0;
    x11_get_screen_size(&width,&height);

if(width>=2560){
#if(QT_VERSION>=QT_VERSION_CHECK(5,6,0))
    QCoreApplication::setAttribute(Qt::AA_EnableHighDpiScaling);
    QCoreApplication::setAttribute(Qt::AA_UseHighDpiPixmaps);
#endif
}

    QApplication a(argc, argv);
    a.setQuitOnLastWindowClosed(true);

    QResource::registerResource("image.qrc");

    QTextCodec *codec = QTextCodec::codecForName("UTF-8");
    QTextCodec::setCodecForLocale(codec);

    XEventMonitor::instance()->start();

    QWidget w;
    DisplaySwitch ds(&w);
    QObject::connect(XEventMonitor::instance(), SIGNAL(keyRelease(QString)),
                     &ds, SLOT(onGlobalKeyRelease(QString)));

    return a.exec();
}


manager/common/xeventmonitor.cpp
/* -*- Mode: C++; indent-tabs-mode: nil; tab-width: 4 -*-
 * -*- coding: utf-8 -*-
 *
 * Copyright (C) 2011 ~ 2017 Deepin, Inc.
 *               2011 ~ 2017 Wang Yong
 * Copyright (C) 2018 Tianjin KYLIN Information Technology Co., Ltd.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "xeventmonitor.h"
#include <iostream>
#include <QVector>
#include <QSet>
#include <QX11Info>
#include <X11/Xlibint.h>
#include <X11/XKBlib.h>
#include <X11/Xlib.h>
#include <X11/extensions/record.h>
#include <X11/keysym.h>

// Virtual button codes that are not defined by X11.
#define Button1            1
#define Button2            2
#define Button3            3
#define WheelUp            4
#define WheelDown          5
#define WheelLeft          6
#define WheelRight         7
#define XButton1           8
#define XButton2           9

XEventMonitor *XEventMonitor::instance_ = new XEventMonitor();

QVector<KeySym> ModifiersVec{
    XK_Control_L,
    XK_Control_R,
    XK_Shift_L,
    XK_Shift_R,
    XK_Super_L,
    XK_Super_R,
    XK_Alt_L,
    XK_Alt_R
};

class XEventMonitorPrivate
{
public:
    XEventMonitorPrivate(XEventMonitor *parent);
    virtual ~XEventMonitorPrivate();
    void run();

protected:
    XEventMonitor *q_ptr;
    QSet<KeySym> modifiers;

    bool filterWheelEvent(int detail);
    static void callback(XPointer trash, XRecordInterceptData* data);
    void handleRecordEvent(XRecordInterceptData *);
    void emitButtonSignal(const char *member, xEvent *event);
    void emitKeySignal(const char *member, xEvent *event);
    void updateModifier(xEvent *event, bool isAdd);

private:
    Q_DECLARE_PUBLIC(XEventMonitor)
};

XEventMonitorPrivate::XEventMonitorPrivate(XEventMonitor *parent)
    : q_ptr(parent)
{

}

XEventMonitorPrivate::~XEventMonitorPrivate()
{

}

void XEventMonitorPrivate::emitButtonSignal(const char *member, xEvent *event)
{
    //qDebug() << event->u.u.detail;;

    int x = event->u.keyButtonPointer.rootX;
    int y = event->u.keyButtonPointer.rootY;
    QMetaObject::invokeMethod(q_ptr, member,
                              Qt::DirectConnection,
                              Q_ARG(int, x),
                              Q_ARG(int, y));
}

void XEventMonitorPrivate::emitKeySignal(const char *member, xEvent *event)
{
    Display *display = XOpenDisplay(NULL);
    int keyCode = event->u.u.detail;
    KeySym keySym = XkbKeycodeToKeysym(display, event->u.u.detail, 0, 0);

    QString keyStrSplice;
    for(auto modifier : modifiers)
    {
        keyStrSplice += QString(XKeysymToString(modifier)) + "+";
    }
    //按键是修饰键
    if(ModifiersVec.contains(keySym) && !modifiers.isEmpty())
        keyStrSplice.remove(keyStrSplice.length() - 1, 1);
    else
        keyStrSplice += XKeysymToString(keySym);

    QMetaObject::invokeMethod(q_ptr, member,
                              Qt::AutoConnection,
                              Q_ARG(int, keyCode));
    QMetaObject::invokeMethod(q_ptr, member,
                              Qt::AutoConnection,
                              Q_ARG(QString, keyStrSplice));
    XCloseDisplay(display);
}

void XEventMonitorPrivate::run()
{
    Display* display = XOpenDisplay(0);
    if (display == 0) {
        fprintf(stderr, "unable to open display\n");
        return;
    }

    // Receive from ALL clients, including future clients.
    XRecordClientSpec clients = XRecordAllClients;
    XRecordRange* range = XRecordAllocRange();
    if (range == 0) {
        fprintf(stderr, "unable to allocate XRecordRange\n");
        return;
    }

    // Receive KeyPress, KeyRelease, ButtonPress, ButtonRelease and MotionNotify events.
    memset(range, 0, sizeof(XRecordRange));
    range->device_events.first = KeyPress;
    range->device_events.last  = MotionNotify;

    // And create the XRECORD context.
    XRecordContext context = XRecordCreateContext(display, 0, &clients, 1, &range, 1);
    if (context == 0) {
        fprintf(stderr, "XRecordCreateContext failed\n");
        return;
    }
    XFree(range);

    XSync(display, True);

    Display* display_datalink = XOpenDisplay(0);
    if (display_datalink == 0) {
        fprintf(stderr, "unable to open second display\n");
        return;
    }
    if (!XRecordEnableContext(display_datalink, context,  callback, (XPointer) this)) {
        fprintf(stderr, "XRecordEnableContext() failed\n");
        return;
    }

}

void XEventMonitorPrivate::callback(XPointer ptr, XRecordInterceptData* data)
{
    ((XEventMonitorPrivate *) ptr)->handleRecordEvent(data);
}

void XEventMonitorPrivate::handleRecordEvent(XRecordInterceptData* data)
{

    if (data->category == XRecordFromServer) {
        xEvent * event = (xEvent *)data->data;

        switch (event->u.u.type)
        {
        case ButtonPress:
            if (filterWheelEvent(event->u.u.detail)) {
                emitButtonSignal("buttonPress", event);
            }
            break;
        case MotionNotify:
            emitButtonSignal("buttonDrag", event);
            break;
        case ButtonRelease:
            if (filterWheelEvent(event->u.u.detail)) {
                emitButtonSignal("buttonRelease", event);
            }
            break;
        case KeyPress:
            updateModifier(event, true);
            emitKeySignal("keyPress", event);
            break;
        case KeyRelease:
            updateModifier(event, false);
            emitKeySignal("keyRelease", event);
            break;
        default:
            break;
        }
    }
    fflush(stdout);
    XRecordFreeData(data);
}

bool XEventMonitorPrivate::filterWheelEvent(int detail)
{
    return detail != WheelUp && detail != WheelDown && detail != WheelLeft && detail != WheelRight;
}

void XEventMonitorPrivate::updateModifier(xEvent *event, bool isAdd)
{
    Display *display = XOpenDisplay(NULL);
    KeySym keySym = XkbKeycodeToKeysym(display, event->u.u.detail, 0, 0);

    if(ModifiersVec.contains(keySym))
    {
        if(isAdd)
        {
            modifiers.insert(keySym);
        }
        else
        {
            modifiers.remove(keySym);
        }
    }

    XCloseDisplay(display);
}

XEventMonitor::XEventMonitor(QObject *parent)
    : QThread(parent),
      d_ptr(new XEventMonitorPrivate(this))
{
    Q_D(XEventMonitor);
}

XEventMonitor::~XEventMonitor()
{
    requestInterruption();
    quit();
    wait();
}

void XEventMonitor::run()
{
    if(!isInterruptionRequested())
    {
        d_ptr->run();
    }
}

bool checkCapsState()
{
    //判断大写键状态
    bool capsState = false;
    unsigned int n;
    XkbGetIndicatorState(QX11Info::display(), XkbUseCoreKbd, &n);
    capsState = (n & 0x01) == 1;
	return capsState;
}

bool checkNumLockState()
{
    XKeyboardState x;
    XGetKeyboardControl(QX11Info::display(), &x);
    return x.led_mask & 2;
}


manager/display-switch/displayswitch.h
/* displayswitch.h
 * Copyright (C) 2018 Tianjin KYLIN Information Technology Co., Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301, USA.
**/
#ifndef DISPLAYSWITCH_H
#define DISPLAYSWITCH_H

#include <QMainWindow>

namespace Ui {
class MainWindow;
}

class DisplaySwitch : public QMainWindow
{
    Q_OBJECT
public:
    explicit DisplaySwitch(QWidget *parent = nullptr);
    void initUi();
    void setQSS();

signals:

public slots:
    void onDisplaySwitchSelect();
    void onDisplaySwitchConfirm();
    void onPositionChanged(const QRect&);
    void onGlobalKeyRelease(const QString &key);

private:
    Ui::MainWindow *ui;
    int     currentMode;
    bool    selecting;
};

#endif // DISPLAYSWITCH_H


manager/display-switch/displayservice.h
/* displayservice.h
 * Copyright (C) 2018 Tianjin KYLIN Information Technology Co., Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301, USA.
**/
#ifndef DISPLAYSERVICE_H
#define DISPLAYSERVICE_H

#include <QObject>
#include <QVector>
#include <QMap>

enum DisplayMode{
    DISPLAY_MODE_ORI        = 1,
    DISPLAY_MODE_CLONE      = 2,
    DISPLAY_MODE_EXTEND     = 3,
    DISPLAY_MODE_ONLY_OUT   = 4
};

class DisplayService : public QObject
{
    Q_OBJECT
public:
    explicit DisplayService(QObject *parent = nullptr);
    bool switchDisplayMode(DisplayMode targetMode);

private:
    // 屏幕的刷新频率
    class  DispResolution
    {
    public:
         DispResolution() {}
         DispResolution(int w, int h, double r): width(w), height(h), flushHz(r){}
    public:
         int width, height;
         double flushHz;
    };

    void getMonitors();
    // flag = true, 从index正向找到第一个正整数
    // flag = false, 从index反向找到第一个正整数
    // 没有找到则返回 0
    QString getNumberFromStr(const QString &str, int index, const bool flag);
    QString createSuitableResolution(); // clone模式下 当没有适合各个屏幕的分辨率时  就创造一个最合适的分辨率

    QMap<QString, QVector<QString>> monitors;
    QStringList monitorNames;
};

#endif // DISPLAYSERVICE_H


manager/display-switch/monitorwatcher.h
/* monitorwatcher.h
 * Copyright (C) 2018 Tianjin KYLIN Information Technology Co., Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301, USA.
**/
#ifndef MONITORWATCHER_H
#define MONITORWATCHER_H

#include <QThread>
#include <QMap>
#include <QSize>
#include <QVector>

/*!
 * \brief The MonitorWatcher class
 * 监控显示器的插拔
 */
class MonitorWatcher : public QThread
{
    Q_OBJECT

public:
    static MonitorWatcher* instance(QObject* parent=nullptr);
    ~MonitorWatcher();
    QSize getVirtualSize();
    int getMonitorCount();

signals:
    void monitorCountChanged(int newCount);
    void virtualSizeChanged(const QSize &newVirtualSize);

protected:
    void run();

private:
    MonitorWatcher(QObject *parent=nullptr);
    QSize getMonitorMaxSize(const QString &drm);

    void getMonitors();

    static MonitorWatcher* _instance;

    QMap<QString, QVector<QString>> monitors;
    QStringList monitorNames;

    QMap<QString, QString>  drmStatus;
    QSize                   virtualSize;
    int                     monitorCount;
};

#endif // MONITORWATCHER_H


manager/display-switch/displayswitch.cpp
/* displayswitch.cpp
 * Copyright (C) 2018 Tianjin KYLIN Information Technology Co., Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301, USA.
**/
#include "displayswitch.h"
#include "ui_mainwindow.h"
#include <QFile>
#include <QDebug>
#include <QDesktopWidget>
#include "displayservice.h"

DisplaySwitch::DisplaySwitch(QWidget *parent)
    : QMainWindow(parent),
      ui(new Ui::MainWindow()),
      currentMode(0),
      selecting(false)
{
    ui->setupUi(this);
    this->setAttribute(Qt::WA_TranslucentBackground);
    setWindowFlags(Qt::FramelessWindowHint | Qt::ToolTip);
    initUi();
    setQSS();

}

void DisplaySwitch::initUi()
{
    ui->centralWidget->setAutoFillBackground(true);
    QPalette pc;
    QPixmap img_bg = QPixmap(":/res/bg.png");
    pc.setBrush(QPalette::Window, QBrush(img_bg));
    ui->centralWidget->setPalette(pc);

    ui->lb_mode_1->setText("仅计算机");
    ui->lb_mode_2->setText("复制");
    ui->lb_mode_3->setText("扩展");
    ui->lb_mode_4->setText("仅投影仪");

    ui->mode_1->setFocusPolicy(Qt::NoFocus);
    ui->mode_2->setFocusPolicy(Qt::NoFocus);
    ui->mode_3->setFocusPolicy(Qt::NoFocus);
    ui->mode_4->setFocusPolicy(Qt::NoFocus);

    QDesktopWidget *dw = QApplication::desktop();
    QRect rect = dw->screen(dw->primaryScreen())->geometry();
    setGeometry(rect.left() + (rect.width() - width()) / 2,
                rect.top() + (rect.height() - height()) / 2,
                width(), height());
}

void DisplaySwitch::setQSS()
{
    QString style_sheet = "QLabel#bg_mode_1{"
                        "    background-image:url(:res/btn1.png);"
                        "}"
                        "QLabel#bg_mode_2{"
                        "    background-image:url(:res/btn2.png);"
                        "}"
                        "QLabel#bg_mode_3{"
                        "    background-image:url(:res/btn3.png);"
                        "}"
                        "QLabel#bg_mode_4{"
                        "    background-image:url(:res/btn4.png);"
                        "}"
                        "QLabel#lb_mode{"
                        "    color:black;"
                        "    font-family:'droid-fallback';"
                        "}"
                        "QLabel#split{"
                        "    background-image:url(:res/split.png);"
                        "    border:0px;"
                        "}"
                        "QPushButton#mode{"
                        "    background-image:url(:res/hover.png);"
                        "    border:0px;"
                        "}";
        setStyleSheet(style_sheet);
}

void DisplaySwitch::onDisplaySwitchSelect(/*const QRect &rect*/)
{
    show();

    ui->mode_1->hide();
    ui->mode_2->hide();
    ui->mode_3->hide();
    ui->mode_4->hide();
    currentMode++;
    if(currentMode > 4)
        currentMode = 1;
    switch(currentMode) {
    case 1:
        ui->mode_1->show();
        break;
    case 2:
        ui->mode_2->show();
        break;
    case 3:
        ui->mode_3->show();
        break;
    case 4:
        ui->mode_4->show();
        break;
    }
}

void DisplaySwitch::onDisplaySwitchConfirm()
{
    if(!isHidden()) {
        hide();
        DisplayService().switchDisplayMode(DisplayMode(currentMode));
        currentMode--;
    }
}

void DisplaySwitch::onPositionChanged(const QRect &rect)
{
    move(rect.left() + (rect.width() - width()) / 2,
               rect.top() + (rect.height() - height()) / 2);
}

void DisplaySwitch::onGlobalKeyRelease(const QString &key)
{
    if(key == "Super_L+p" || key == "Super_L+F3")
    {
        selecting = true;
        onDisplaySwitchSelect();
    }
    else if(key == "Super_L")
    {
        if(selecting)
        {
            onDisplaySwitchConfirm();
            selecting =  false;
        }
    }
}


manager/display-switch/displayservice.cpp
/* displayservice.cpp
 * Copyright (C) 2018 Tianjin KYLIN Information Technology Co., Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301, USA.
**/
#include "displayservice.h"
#include "monitorwatcher.h"
#include <QException>
#include <QDebug>
#include <QProcess>
#include <QStandardPaths>
#include <QVector>
#include <QtCore/qmath.h>

DisplayService::DisplayService(QObject *parent) : QObject(parent)
{

}

bool DisplayService::switchDisplayMode(DisplayMode targetMode)
{
    //获取显示器列表
    getMonitors();

    //如果当前只有一个显示器就什么也不做
    if(monitors.keys().size() < 2) {
        return false;
    }

    QProcess subProcess;
    //模式切换
    try {
        switch(targetMode){
        case DISPLAY_MODE_ORI:
        {
            QString XRANDR_ORIONLY = "xrandr --output " + monitorNames[0] + " --primary --auto";
            for(int i = 1; i < monitorNames.size(); i++)
                XRANDR_ORIONLY = XRANDR_ORIONLY + " --output " + monitorNames[i] + " --off";

            qDebug() << "info: [Class:DisplayService][switchDisplayMode] XRANDR_ORIONLY: " << XRANDR_ORIONLY;
            subProcess.start(XRANDR_ORIONLY, QIODevice::NotOpen);
            break;
        }
        case DISPLAY_MODE_CLONE:
        {
            //查找最佳克隆分辨率
            QString BEST_CLONE_MODE = createSuitableResolution();
            /*
            bool find = false;
            for(auto & mode_0 : monitors[monitorNames[0]]) {
                for(auto & mode_1 : monitors[monitorNames[1]]) {
                    if(mode_0 == mode_1) {
                        BEST_CLONE_MODE = mode_0;
                        find = true;
                        break;
                    }
                }
                if(find)
                    break;
            }
            */
            QString XRANDR_CLONE = "xrandr --output " + monitorNames[0] + " --mode " + BEST_CLONE_MODE + "  --primary --auto ";
            for(int i = 1; i < monitorNames.size(); i++)
                XRANDR_CLONE = XRANDR_CLONE + " --output " + monitorNames[i] + " --mode " + BEST_CLONE_MODE + " --same-as " + monitorNames[0];

             qDebug() << "info: [Class:DisplayService][switchDisplayMode] XRANDR_CLONE: " << XRANDR_CLONE;
             subProcess.start(XRANDR_CLONE, QIODevice::NotOpen);
            break;
        }
        case DISPLAY_MODE_EXTEND:
        {
            QString XRANDR_EXTEND = "xrandr --output " + monitorNames[0] + "  --primary --auto";
            for(int i = 1; i < monitorNames.size(); i++)
                XRANDR_EXTEND = XRANDR_EXTEND + " --output " + monitorNames[i] + " --right-of " + monitorNames[i-1] + " --auto";

            qDebug() << "info: [Class:DisplayService][switchDisplayMode] XRANDR_EXTEND: " << XRANDR_EXTEND;
            subProcess.start(XRANDR_EXTEND, QIODevice::NotOpen);
            break;
        }
        case DISPLAY_MODE_ONLY_OUT:
        {
            QString XRANDR_OUTONLY = "xrandr --output " + monitorNames[0] + " --off --output " + monitorNames[1] + " --primary --auto";

            subProcess.start(XRANDR_OUTONLY);
            qDebug() << "info: [Class:DisplayService][switchDisplayMode] XRANDR_OUTONLY: " << XRANDR_OUTONLY;
            break;
        }
        }
        subProcess.waitForFinished();
        return true;
    } catch(const QException &e) {
        qWarning() << "info: [Class:DisplayService][switchDisplayMode]: " << e.what();
        return false;
    }
}


void DisplayService::getMonitors()
{
    QProcess subProcess;
    subProcess.setProgram(QStandardPaths::findExecutable("xrandr"));
    subProcess.setArguments({"-q"});
    subProcess.start(QIODevice::ReadOnly);
    subProcess.waitForFinished();
    QString outputs = subProcess.readAll();
    QStringList lines = outputs.split('\n');
    QString name;
    QVector<QString> modes;
    bool find = false;
    QString lastName;
    monitorNames.clear();
    monitors.clear();
    for(auto & line : lines) {
        if(line.indexOf(" connected") != -1) {   //找显示器名称
            name = line.left(line.indexOf(' '));
            monitorNames.push_back(name);
            if(find)    //又找到了一个显示器，将上一个显示器的信息存到map
                monitors[lastName] = modes;
            find = true;
            lastName = name;
            modes.clear();
        } else {
            if(line.startsWith(' ')) {      //获取分辨率
                QString mode = line.trimmed().split(' ').at(0);
                modes.push_back(mode);
            }
        }
    }
    monitors[name] = modes;     //将最后一个显示器的信息存储到map

}

QString DisplayService::getNumberFromStr(const QString &str, int index, const bool flag)
{
    int dir = flag ? 1 : -1;
    const int &len = str.size();
    QString resStr = "";
    while(index < len && index >= 0 && (str.at(index) > '9' || str.at(index) < '0')) index += dir;
    while(index < len && index >= 0 && str.at(index) <= '9' && str.at(index) >= '0')
    {
        resStr = flag ? resStr + str.at(index) : str.at(index) + resStr;
        index += dir;
    }
    if(resStr.isEmpty())
        return "0";
    return resStr;
}

QString DisplayService::createSuitableResolution()
{
    QVector<DispResolution> dispsVec; // 记录各个屏幕的最适合的分辨率

    QString cmd = "xrandr -q";
    QProcess subProcess;
    subProcess.start(cmd, QIODevice::ReadOnly);
    subProcess.waitForFinished();
    QString outputs = subProcess.readAll();
    QStringList lines = outputs.split('\n');


    DispResolution minDisp(INT32_MAX, INT32_MAX, 60);
    for(int i = 0; i < lines.size();++i)
    {
        const QString &line = lines.at(i);
        if(line.indexOf(" connected") != -1)
        {
            const QString &nextLine = lines.at(++i);
            int x_index = nextLine.indexOf("x");
            int p_index = nextLine.indexOf(".");
            int width = getNumberFromStr(nextLine, x_index, false).toInt();
            int height = getNumberFromStr(nextLine, x_index, true).toInt();
            QString s1 = getNumberFromStr(nextLine, p_index, false);
            QString s2 = getNumberFromStr(nextLine, p_index, true);
            double flushHz = s1.toInt() + s2.toInt() / pow(10, s2.size());
            dispsVec.push_back({width, height, flushHz});
            //qDebug() << "获得第 " << dispsVec.size() << " 个屏幕的最优分辨率: "\
                     << width << " " << height << " " << flushHz;
            minDisp.width = minDisp.width < width ? minDisp.width : width;
            minDisp.height = minDisp.height < height ? minDisp.height : height;
            minDisp.flushHz = minDisp.flushHz < flushHz ? minDisp.flushHz : flushHz;
        }
    }

    //qDebug() << "生成最优分辨率: " << minDisp.width << " " << minDisp.height << " " << minDisp.flushHz;

    cmd = "cvt " + QString::number(minDisp.width) + " "
            + QString::number(minDisp.height) + " "
            + QString::number(minDisp.flushHz);

    subProcess.start(cmd, QIODevice::ReadOnly);
    subProcess.waitForFinished();
    outputs = subProcess.readAll();

    QString resolutionName;
    int begin_index = outputs.indexOf("Modeline ") + 8;
    cmd = "xrandr --newmode " + outputs.mid(begin_index);
    //qDebug() << cmd;
    subProcess.start(cmd, QIODevice::NotOpen);
    subProcess.waitForFinished();

    begin_index = outputs.indexOf("\"") + 1;
    int end_index = outputs.indexOf("\"", begin_index);
    resolutionName = outputs.mid(begin_index, end_index - begin_index);
    //qDebug() << "resolutionName: " << resolutionName;

    getMonitors();
    for(int i = 0;i < monitorNames.size();i++)
    {
        cmd = "xrandr --addmode " + monitorNames.at(i) + " " + resolutionName;
        //qDebug() << cmd;
        subProcess.start(cmd, QIODevice::NotOpen);
        subProcess.waitForFinished();
    }

    return resolutionName;
}
